### PR TITLE
[SPARK-34558][SQL][FOLLOWUP] Use final Hadoop conf to instantiate FileSystem in SharedState

### PR DESCRIPTION
**Pull request by cloud-fan**
### What changes were proposed in this pull request?

This is a follow-up of https://github.com/apache/spark/pull/31671

https://github.com/apache/spark/pull/31671 has an unexpected behavior change that it uses a different Hadoop conf (`sparkContext.hadoopConfiguration`) to instantiate `FileSystem`, which is used to qualify the warehouse path. Before https://github.com/apache/spark/pull/31671 , the Hadoop conf to instantiate `FileSystem` is `session.sessionState.newHadoopConf()`.

More specifically, `session.sessionState.newHadoopConf()` has more conf entries:
1. it includes configs from `SharedState.initialConfigs`
2. in includes configs from `sparkContext.conf`

This PR updates `SharedState` to use the final Hadoop conf to instantiate `FileSystem`.

### Why are the changes needed?

fix behavior change

### Does this PR introduce _any_ user-facing change?

yes, the behavior will be the same before https://github.com/apache/spark/pull/31671

### How was this patch tested?

manually check the log of `FileSystem` and verify the passed in configs.

### CODE REVIEW

**Comment by cloud-fan** [comment]
cc @yaooqinn

File: sql/core/src/main/scala/org/apache/spark/sql/internal/SharedState.scala [hunk 1]
```diff
@@ -53,24 +53,33 @@ private[sql] class SharedState(
...
-    // Load hive-site.xml into hadoopConf and determine the warehouse path which will be set into
-    // both spark conf and hadoop conf avoiding be affected by any SparkSession level options
-    val initialConfigsWithoutWarehouse = SharedState.resolveWarehousePath(
+    val warehousePath = SharedState.resolveWarehousePath(
       sparkContext.conf, sparkContext.hadoopConfiguration, initialConfigs)
 
     val confClone = sparkContext.conf.clone()
     val hadoopConfClone = new Configuration(sparkContext.hadoopConfiguration)
+    // Extract entries from `SparkConf` and put them in the Hadoop conf.
+    confClone.getAll.foreach { case (k, v) =>
```
Comment (yaooqinn) [inline sql/core/src/main/scala/org/apache/spark/sql/internal/SharedState.scala:62]: I don't quite get this change. IIUC, only `spark.hadoop` and `spark.hive` prefixed ones get resolved and extracted into Hadoop configuration. After this, it seems that a `hadoop.xyz` has higher priority than `spark.hadoop.hadoop.xyz`.

Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/internal/SharedState.scala:62]: This is to follow `session.sessionState.newHadoopConf()`:
1. The SQLConf contains all the configs from SparkConf, see https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/internal/BaseSessionStateBuilder.scala#L89
2. The Hadoop conf contains all the configs from SQLConf, see https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/internal/SessionState.scala#L127

**Comment by yaooqinn** [comment]
Not related to this pr, a `FileSystem` gets instantiated if the event log is on during `sc` instantiating, with `sc.hadoopConfiguration`. Is this also relevant?

File: sql/core/src/main/scala/org/apache/spark/sql/internal/SharedState.scala (hunk 1 above)
Comment (yaooqinn) [reply sql/core/src/main/scala/org/apache/spark/sql/internal/SharedState.scala:62]: Make sense to me

**Review by yaooqinn — APPROVED** [review]

**Comment by cloud-fan** [comment]
It's not related. This PR is to make sure qualifying the warehouse path is the same as before. `FileSystem` may have different qualifying behavior with different configs.

**Comment by cloud-fan** [comment]
thanks for the review, merging to master!

**Pull request closed without merging**